### PR TITLE
Feature: Carbon DeFi

### PR DESCRIPTION
**Pull request by tiagofilipenunes**

### CODE REVIEW

File: src/adapters/carbon-defi/contracts/CarbonController.ts [hunk 1]
```diff
@@ -0,0 +1,1595 @@
...
+import type {
+  TypedContractEvent,
+  TypedDeferredTopicFilter,
+  TypedEventLog,
+  TypedLogDescription,
+  TypedListener,
+  TypedContractMethod,
+} from "./common";
+
+export type OrderStruct = {
```
Comment (zavelevsky) [inline src/adapters/carbon-defi/contracts/CarbonController.ts:26]: I feel like you are declaring types that already exist and you can simply re-export

File: src/adapters/carbon-defi/products/strategies/carbonDeFiStrategiesAdapter.ts [hunk 2]
```diff
@@ -0,0 +1,509 @@
...
+
+  /**
+   * Update me.
+   * Add logic to get userAddress positions in your protocol
+   */
+  async getPositions({
+    userAddress,
+    blockNumber,
+  }: GetPositionsInput): Promise<ProtocolPosition[]> {
+    const carbonControllerContract = CarbonController__factory.connect(
```
Comment (zavelevsky) [inline src/adapters/carbon-defi/products/strategies/carbonDeFiStrategiesAdapter.ts:142]: can't we pre-define these?

File: src/adapters/carbon-defi/products/strategies/carbonDeFiStrategiesAdapter.ts [hunk 3]
```diff
@@ -0,0 +1,509 @@
...
+      contractAddresses[this.chainId]!.carbonControllerAddress,
+      this.provider,
+    )
+    const voucherContract = Voucher__factory.connect(
+      contractAddresses[this.chainId]!.voucherContractAddress,
+      this.provider,
+    )
+
+    const strategyIds: BigNumberish[] = await voucherContract.tokensByOwner(
```
Comment (zavelevsky) [inline src/adapters/carbon-defi/products/strategies/carbonDeFiStrategiesAdapter.ts:151]: SDK has getUserStrategies - not relevant?

Comment (tiagofilipenunes) [reply src/adapters/carbon-defi/products/strategies/carbonDeFiStrategiesAdapter.ts:151]: As per their Contributing instructions:
https://github.com/bancorprotocol/mmi-defi-adapters/blob/9a4e279f493be26aa4a1cd9242348d852fca1020/CONTRIBUTING.md?plain=1#L26

File: src/adapters/carbon-defi/products/strategies/carbonDeFiStrategiesAdapter.ts (hunk 2 above)
Comment (tiagofilipenunes) [reply src/adapters/carbon-defi/products/strategies/carbonDeFiStrategiesAdapter.ts:142]: It's auto-generated using `npm run build-types` - was useful in getting types for event logs. The approach is part of the acceptance criteria in the readme:
https://github.com/bancorprotocol/mmi-defi-adapters/blob/9a4e279f493be26aa4a1cd9242348d852fca1020/README.md?plain=1#L79-L90

**Pull request closed without merging**